Bluetooth: controller: Fix ticker to avoid recursive ticker_job

cvinayak opened the pull request:
Fix ticker job to defer itself to avoid recursive
invocation to itself due to ticker interface calls from
inside the ticker operation callbacks.

The recursive use was exposed when using ticker stop
operation callback of stopping an auxiliary PDU to stop
the primary PDU scheduling as part of generation of
Advertising Terminate event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix ticker to avoid recursive ticker_job

Fix ticker job to defer itself to avoid recursive
invocation to itself due to ticker interface calls from
inside the ticker operation callbacks.

The recursive use was exposed when using ticker stop
operation callback of stopping an auxiliary PDU to stop
the primary PDU scheduling as part of generation of
Advertising Terminate event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -2276,11 +2276,18 @@ void ticker_job(void *param)
+		return;
+	}
 	instance->job_guard = 1U;
 
```
Consider also moving clearing of the job_guard to after `ticker_job_compare_update`

cvinayak replied:
Is this because `ticks_slot_previous` and `ticks_current` are updated in `ticker_job_compare_update`, and the vectoring (swap) to `ticker_worker` function would use stale values? or do you have other reasons? 

(sorry for asking again, from the message here: https://github.com/zephyrproject-rtos/zephyr/pull/31026#issue-546079708, I understand context-safety, but just making double sure :-) ).

mtpr-ot replied:
> Is this because `ticks_slot_previous` and `ticks_current` are updated in `ticker_job_compare_update`, and the vectoring (swap) to `ticker_worker` function would use stale values? or do you have other reasons?

Yes, that is exactly why. The suggestion is not based on an actual observation, just based on analysis and being cautious.

cvinayak replied:
Agreed. Your analysis is correct, as the resolution happens in ticker_worker, there are chances that after a current reserving ticker is stopped, it is necessary to have correct value of `ticks_slot_previous` and `ticks_current` be used in the ticker_worker.

cvinayak committed:
Bluetooth: controller: Fix ticker job_guard

Move the ticker job_guard reset to after
ticker_job_compare_update, so that the ticks_current and
ticks_slot_previous are updated before ticker_worker gets
to execute. Without this fix, there is a possibility that
ticker_worker will use incorrect ticks_slot_previous and
ticks_current value under race conditions.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

mtpr-ot reviewed (approved):
Looks good!

kruithofa reviewed (approved):
LGTM
